mobile/src/main/java/ch/epfl/sdp/mobile/ui/play/ExpandableFloatingActionButton.kt

```diff
@@ -0,0 +1,240 @@
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.geometry.Size
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.unit.Density
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.util.lerp
+import ch.epfl.sdp.mobile.ui.FadeEnterTransition
+import ch.epfl.sdp.mobile.ui.FadeExitTransition
+import ch.epfl.sdp.mobile.ui.PawniesColors.Beige050
+import ch.epfl.sdp.mobile.ui.PawniesColors.Green100
+import ch.epfl.sdp.mobile.ui.PawniesColors.Green800
+
+/** The state of an [ExpandableFloatingActionButton] */
+interface ExpandableFloatingActionButtonState {
+
+  /** True iff the button should be currently expanded. */
+  var expanded: Boolean
+}
+
+/**
+ * An interface representing the colors of an [ExpandableFloatingActionButton]. The colors may
+ * change depending on whether the button is currently expanded or not.
+ */
+interface ExpandableFloatingActionButtonColors {
+
+  /**
+   * Returns a state with the color for the button background.
+   *
+   * @param expanded true if the button is currently expanded.
+   */
+  @Composable fun background(expanded: Boolean): State<Color>
+
+  /**
+   * Returns a state with the color for the button border.
+   *
+   * @param expanded true if the button is currently expanded.
+   */
+  @Composable fun border(expanded: Boolean): State<Color>
+
+  /**
+   * Returns a state with the color for the button content color.
+   *
+   * @param expanded true if the button is currently expanded.
+   */
+  @Composable fun contentColor(expanded: Boolean): State<Color>
+}
+
+/**
+ * A composable which displays a floating action which may be expanded.
+ *
+ * @param expandedContent the content to display when expanded. Usually a list of
+ * [ExpandableFloatingActionButtonItem].
+ * @param modifier the [Modifier] for this composable.
+ * @param state the backing state for this button.
+ * @param colors the colors for this button.
+ * @param elevation the elevation for this button.
+ * @param interactionSource the [MutableInteractionSource] to keep track of inputs.
+ * @param expandedCornerSize the size of the rounded corners when the button is expanded.
+ * @param collapsedContent the content to display when collapsed.
+ */
+@OptIn(
+    ExperimentalAnimationApi::class,
+    ExperimentalMaterialApi::class,
+)
+@Composable
+fun ExpandableFloatingActionButton(
+    expandedContent: @Composable ColumnScope.() -> Unit,
+    modifier: Modifier = Modifier,
+    state: ExpandableFloatingActionButtonState = rememberExpandableFloatingActionButtonState(),
+    colors: ExpandableFloatingActionButtonColors = DefaultExpandableFloatingActionButtonColors,
+    elevation: ButtonElevation = ButtonDefaults.elevation(),
+    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
+    expandedCornerSize: Dp = 8.dp,
+    collapsedContent: @Composable RowScope.() -> Unit,
```

Comment:
Question: What does `RowScope.() -> Unit` mean. Why do you need the RowScope part ?

Reply:
From my understanding` RowScope.() -> Unit` is a Kotlin way to define function types with receivers more [here](https://kotlinlang.org/docs/lambdas.html#function-literals-with-receiver)
If I understood it correctly, the `RowScope` part means that the `collapedContent` composable is drawn from a `RowScope` context.
I'll let @alexandrepiveteau clarify and I hope I am not giving none sense explanation 😁

Reply:
That's exactly right @BadrTad 👍 

`RowScope` and `ColumnScope` typically both provide some additional `Modifier` which you can therefore only use when you're in a `RowScope`. An example is `Modifier.align()` (which would not have a well-defined behaviour in layouts which are not rows or columns), so these scopes are used to provide a "type-safe" DSL for building the contents of some rows or columns.

Reply:
So long story short, the `RowScope().` receiver is not needed. However, if you want to align some items at the top / bottom, you can apply `Modifier.align()` in the `collapsedContent` composable body.